### PR TITLE
[3.3.5] ChatCommands, the other half: chat command resolution refactor

### PR DESCRIPTION
**Pull request by Treeston**
Now that the command handler side is cleaned up with the dark arts of C++17 and template logic, time to tackle the other half of chat commands – the part that takes _what you enter in chat_ and comes up with _which command should be executed_.

Two main things to do here:

* [x] Make command tables more readable in scripts
* [x] Get rid of `strtok` and move everything to `std::string_view` where it should be.

### CODE REVIEW

**Comment by Treeston** [comment]
_(NB: This will not currently build. I have only changed one command script file (`cs_learn`) to the new constructors for the time being.)_

**Comment by Treeston** [comment]
@Carbenium this is relevant to your interests

File: src/server/game/Chat/Chat.cpp [hunk 1]
```diff
@@ -1185,16 +1177,14 @@ int CliHandler::GetSessionDbLocaleIndex() const
...
 
-bool AddonChannelCommandHandler::ParseCommands(char const* str)
+bool AddonChannelCommandHandler::ParseCommands(std::string_view str)
 {
-    if (memcmp(str, "TrinityCore\t", 12))
+    if (str.length() < 17)
         return false;
-    char opcode = str[12];
-    if (!opcode) // str[12] is opcode
+    if (str.substr(0, 12) != "TrinityCore\t")
```
Comment (Warpten) [inline src/server/game/Chat/Chat.cpp:1184]: C++20 starts_with (advstd?)

**Comment by jackpoz** [comment]
is this going to be backward compatible with existing commands and command definitions ? I'm thinking about custom patches which added new commands

**Comment by Treeston** [comment]
The `std::vector<ChatCommand>` will need to be replaced with `ChatCommandTable`. _(And I'll see if there is maybe some way we can work around that.)_ Apart from that, yes.

(However, you will get warnings for using the deprecated constructor.)

**Comment by Treeston** [comment]
there you go @jackpoz, now std::vector<ChatCommand> will still work for legacy patches (with deprecation warnings)

**Comment by Treeston** [comment]
todo port all the command scripts and get rid of the warning pragmas

File: src/server/scripts/Commands/cs_account.cpp [hunk 2]
```diff
@@ -41,59 +41,43 @@ EndScriptData */
...
+            { "delete",             HandleAccountDeleteCommand,         rbac::RBAC_PERM_COMMAND_ACCOUNT_DELETE,             Console::Yes       },
+            { "email",              HandleAccountEmailCommand,          rbac::RBAC_PERM_COMMAND_ACCOUNT_EMAIL,              Console::No       },
+            { "onlinelist",         HandleAccountOnlineListCommand,     rbac::RBAC_PERM_COMMAND_ACCOUNT_ONLINE_LIST,        Console::Yes       },
+            { "lock country",       HandleAccountLockCountryCommand,    rbac::RBAC_PERM_COMMAND_ACCOUNT_LOCK_COUNTRY,       Console::No      },
+            { "lock ip",            HandleAccountLockIpCommand,         rbac::RBAC_PERM_COMMAND_ACCOUNT_LOCK_IP,            Console::No      },
+            { "set",                accountSetCommandTable },
+            { "password",           HandleAccountPasswordCommand,       rbac::RBAC_PERM_COMMAND_ACCOUNT_PASSWORD,           Console::No  },
+            { "",                   HandleAccountCommand,               rbac::RBAC_PERM_COMMAND_ACCOUNT,                    Console::No       },
```
Comment (Shauren) [inline src/server/scripts/Commands/cs_account.cpp:76]: so you realigned everything but left the closing } broken, nice

File: src/server/scripts/Commands/cs_account.cpp [hunk 3]
```diff
@@ -48,48 +48,36 @@ class account_commandscript : public CommandScript
...
+        static ChatCommandTable accountSetCommandTable =
+        {
+            { "addon",              HandleAccountSetAddonCommand,       rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_ADDON,          Console::Yes       },
+            { "sec regmail",        HandleAccountSetRegEmailCommand,    rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_SEC_REGMAIL,    Console::Yes },
+            { "sec email",          HandleAccountSetEmailCommand,       rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_SEC_EMAIL,      Console::Yes },
+            { "gmlevel",            HandleAccountSetSecLevelCommand,    rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_SECLEVEL,       Console::Yes       },  // temp for a transition period
+            { "seclevel",           HandleAccountSetSecLevelCommand,    rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_SECLEVEL,       Console::Yes       },
+            { "password",           HandleAccountSetPasswordCommand,    rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_PASSWORD,       Console::Yes      },
+            { "2fa",                HandleAccountSet2FACommand,         rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_2FA,            Console::Yes      },
```
Comment (Faq) [inline src/server/scripts/Commands/cs_account.cpp:61]:
```suggestion
            { "addon",              HandleAccountSetAddonCommand,       rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_ADDON,          Console::Yes },
            { "sec regmail",        HandleAccountSetRegEmailCommand,    rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_SEC_REGMAIL,    Console::Yes },
            { "sec email",          HandleAccountSetEmailCommand,       rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_SEC_EMAIL,      Console::Yes },
            { "gmlevel",            HandleAccountSetSecLevelCommand,    rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_SECLEVEL,       Console::Yes },  // temp for a transition period
            { "seclevel",           HandleAccountSetSecLevelCommand,    rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_SECLEVEL,       Console::Yes },
            { "password",           HandleAccountSetPasswordCommand,    rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_PASSWORD,       Console::Yes },
            { "2fa",                HandleAccountSet2FACommand,         rbac::RBAC_PERM_COMMAND_ACCOUNT_SET_2FA,            Console::Yes },
```

Comment (Faq) [reply src/server/scripts/Commands/cs_account.cpp:61]: Same for all.

File: src/server/worldserver/CommandLine/CliRunnable.cpp [hunk 4]
```diff
@@ -44,32 +44,11 @@ static inline void PrintCliPrefix()
 #if TRINITY_PLATFORM != TRINITY_PLATFORM_WINDOWS
 char* command_finder(char const* text, int state)
 {
-    static size_t idx, len;
-    char const* ret;
-    std::vector<ChatCommand> const& cmd = ChatHandler::getCommandTable();
+    (void)text;
+    (void)state;
+    //@todo maybe? dear lord this readline business is UGLY
```
Comment (Shauren) [inline src/server/worldserver/CommandLine/CliRunnable.cpp:49]: literally unmergable

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/Chat/ChatCommands/ChatCommand.h [hunk 5]
```diff
@@ -98,63 +115,130 @@ namespace Trinity::Impl::ChatCommands
...
+        {}
+        ChatCommandBuilder(ChatCommandBuilder const&) = default;
+
+        template <typename TypedHandler>
+        [[deprecated]] ChatCommandBuilder(char const* name, rbac::RBACPermissions permission, bool console, TypedHandler* handler, char const*)
+            : ChatCommandBuilder(name, *handler, permission, console ? Trinity::ChatCommands::Console::Yes : Trinity::ChatCommands::Console::No) {}
+        [[deprecated]] ChatCommandBuilder(char const* name, rbac::RBACPermissions, bool, std::nullptr_t, char const*, std::vector <ChatCommandBuilder> const& sub)
+            : ChatCommandBuilder(name, sub) {}
+
+        private:
```
Comment (Shauren) [inline src/server/game/Chat/ChatCommands/ChatCommand.h:229]: .

**Pull request merged**